### PR TITLE
[Wallet] Making CWalletTx store TransactionRef.

### PR DESCRIPTION
**Pull request by furszy**
Instead of inheriting from `CTransaction`, now `CWalletTx` will store it as member. Another step forward over the transaction complete immutability goal.

An adaptation of c3f5673a6304e3ea9fa56fff66b6ea1cb73cc98f with further needed changes, essentially:

> it makes it actually fully immutable and never modifies any of its elements (apart from wit) after construction.
> 
> To do so, we heavily rely on C++11. CTransaction gets a (CMutableTransaction&&) constructor that efficiently converts a CMutableTransaction into a CTransaction without copying. In addition, CTransaction gets a deserializing constructors.
> 
> One downside is that CWalletTx cannot easily inherent from CTransaction anymore, as CWalletTx needs a way to modify the CTransaction data inside. By turning the superclass into a CTransactionRef field, it can take advantage (not implemented yet) of sharing CTransactions with the mempool.

### CODE REVIEW

**Comment by furszy** [comment]
Rebased on master + completed the work.
Removed the draft tag, ready for review.

**Review by random-zebra — APPROVED** [review]
Minor concern over the newly introduced `IsNormalTx` function. 
Otherwise looking good, utACK 501edb9a579a68d5781e8d14866802ceb1436a79

File: src/primitives/transaction.h [hunk 1]
```diff
@@ -216,8 +216,33 @@ class CTxOut
...
+inline void SerializeTransaction(TxType& tx, Stream& s, Operation ser_action) {
+    READWRITE(*const_cast<int16_t*>(&tx.nVersion));
+    READWRITE(*const_cast<int16_t*>(&tx.nType));
+    READWRITE(*const_cast<std::vector<CTxIn>*>(&tx.vin));
+    READWRITE(*const_cast<std::vector<CTxOut>*>(&tx.vout));
+    READWRITE(*const_cast<uint32_t*>(&tx.nLockTime));
+
+    if (g_IsSaplingActive && tx.isSaplingVersion()) {
+        READWRITE(*const_cast<Optional<SaplingTxData>*>(&tx.sapData));
+        if (!tx.IsNormalTx()) {
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/primitives/transaction.h]: Updated, function renamed to `IsNormalType`.

File: src/primitives/transaction.h [hunk 2]
```diff
@@ -216,8 +216,33 @@ class CTxOut
...
 struct CMutableTransaction;
 
+/**
+ * Transaction serialization format:
+ * - int32_t nVersion
+ * - std::vector<CTxIn> vin
+ * - std::vector<CTxOut> vout
+ * - uint32_t nLockTime
+ * - Optional<SaplingTxData> sapData
+ * - Optional<uint8_t> extraPayload
```
Comment (random-zebra) [inline src/primitives/transaction.h:226]: typo: `extraPayload` is an optional **vector** of chars, not an optional char.

**Review by random-zebra — COMMENTED** [review]
utACK 7ecef9679a0b1d524d3016bae8f9a910ca5bb2ee

File: src/primitives/transaction.h (hunk 2 above)
Comment (furszy) [reply src/primitives/transaction.h]: typo fixed.

**Review by random-zebra — APPROVED** [review]
utACK 652981a18a57caf41fc49da0a44278e1edefad27

**Review by Fuzzbawls — APPROVED** [review]
utACK 652981a18a57caf41fc49da0a44278e1edefad27

**Pull request merged**